src/storage-client/src/controller.rs

```diff
@@ -2878,6 +2933,180 @@ where
         self.reconcile_managed_collection(id, updates).await;
     }
 
+    /// Partially truncate the statement log tables.
+    /// The logic is as follows:
+    /// For any statement execution whose begin date is younger than
+    /// `statement_logging_retention_time_seconds`, retain both it, and any
+    /// prepared statement and session it references.
+    ///
+    /// Note - this can use a possibly huge amount of memory! If we observe high volume, we might need
+    /// to switch to a more incremental approach, which will require support from Persist, as discussed
+    /// in the comments below.
+    async fn partially_truncate_statement_logging(&mut self) {
+        let now = (self.state.now)();
+        let cutoff = to_datetime(now)
+            - chrono::Duration::seconds(
+                self.state
+                    .config
+                    .statement_logging_retention_time_seconds
+                    .try_into()
+                    .expect("sane config value"),
+            );
+        let mseh_id = self.state.introspection_ids[&IntrospectionType::StatementExecutionHistory];
+        let mpsh_id = self.state.introspection_ids[&IntrospectionType::PreparedStatementHistory];
+        let msh_id = self.state.introspection_ids[&IntrospectionType::SessionHistory];
+        // TODO - If there is a huge number of rows, this can OOM.
+        // We will need to rewrite this to work in an incremental way once Persist
+        // supports that.
+        // See this thread: https://materializeinc.slack.com/archives/C0246GEHL8N/p1690901961209869?thread_ts=1690816977.938789&cid=C0246GEHL8N .
```

Comment:
IMO having an incremental approach here is non-negotiable for this to go to prod. I have https://github.com/MaterializeInc/materialize/pull/20916 up and will polish it up so we can switch to it here

Reply:
When do you think that will be ready? If it will land soon, I'm fine waiting. If it will be a while, I'd rather land this PR now, though; I don't envision it being a problem since we can just avoid turning the `STATEMENT_LOGGING_MAX_SAMPLE_RATE` too high in prod. What do you think?

Reply:
We can likely get it in this week, so long as your use case here doesn't need to read in consolidated state (I haven't read through the implementation). If we aren't able to get it in, let's add a knob to optionally disable truncation, just so we don't wind up in some scenario where we're crash looping because we're reading everything in

Reply:
>  so long as your use case here doesn't need to read in consolidated state

It does, but I think it can be made to avoid that.

> If we aren't able to get it in, let's add a knob to optionally disable truncation

That works for me

Reply:
the persist PR has been merged! let us know how it goes